src/charm.py

```diff
@@ -0,0 +1,323 @@
+            event.extensions = ["hstore:public", "pg_trgm:public"]
+        elif event.database != DATABASE_NAME:
+            # Leader has not yet set requirements. Defer, in case this unit
+            # becomes leader and needs to perform that operation.
+            event.defer()
+
+    # pgsql.MasterChangedEvent is actually defined
+    def _on_master_changed(self, event: pgsql.MasterChangedEvent) -> None:
+        """Handle changes in the primary database unit.
+
+        Args:
+            event: Event triggering the database master changed handler.
+        """
+        if event.master is None:
+            self._stored.db_name = None
+            self._stored.db_user = None
+            self._stored.db_password = None
+            self._stored.db_host = None
+            self._stored.db_port = None
+        else:
+            self._stored.db_name = event.master.dbname
+            self._stored.db_user = event.master.user
+            self._stored.db_password = event.master.password
+            self._stored.db_host = event.master.host
+            self._stored.db_port = event.master.port
+
+        self._on_config_changed(event)
+
+    def _on_install(self, _):
+        """Generate a config template to be rendered later."""
+        self._run_generate_synapse()
+
+    @property
+    def _external_hostname(self):
+        """Return the external hostname passed to ingress via the relation."""
+        # It is recommended to default to `self.app.name` so that the external
+        # hostname will correspond to the deployed application name in the
+        # model, but allow it to be set to something specific via config.
+        return self.config["server_name"] or self.app.name
+
+    def _container(self):
+        """Get the Synapse workload container.
+
+        Returns:
+            The pebble instance of the Synapse container.
+        """
+        return self.unit.get_container(self._CONTAINER_NAME)
+
+    def _populate_synapse_env_settings(self):
+        """Populate env settings to pass around."""
+        if os.getenv("SYNAPSE_SERVER_NAME") is None:
+            os.environ["SYNAPSE_SERVER_NAME"] = self.config["server_name"]
+        if os.getenv("SYNAPSE_REPORT_STATS") is None:
+            os.environ["SYNAPSE_REPORT_STATS"] = self.config["report_stats"]
+        pod_config = {
+                "SYNAPSE_SERVER_NAME": os.getenv("SYNAPSE_SERVER_NAME"),
+                "SYNAPSE_REPORT_STATS": os.getenv("SYNAPSE_REPORT_STATS"),
+                "SYNAPSE_NO_TLS": "True",
+                "POSTGRES_DB": self._stored.db_name,
+                "POSTGRES_HOST": self._stored.db_host,
+                "POSTGRES_PORT": self._stored.db_port,
+                "POSTGRES_USER": self._stored.db_user,
+                "POSTGRES_PASSWORD": self._stored.db_password,
+                }
+        return pod_config
+
+    def _register_user(self, event):
+        """Registers a user for usage with Synapse."""
+        Result = collections.namedtuple(
```

Comment:
Should this be defined in a separate class rather than on each function invocation?

Reply:
Could be, I'd defer it to later though.